src/coprocessor/trie/mod.rs

```diff
@@ -0,0 +1,778 @@
+    ) -> bool {
+        let old_value = old_value.unwrap_or_else(|| F::zero());
+
+        let old_verified = self.old_proof.verify(old_root, key, old_value, hash_cache);
+
+        // This is purely an evaluation-time optimization. Don't try to reproduce in the circuit, which cannot shortcut.
+        if !old_verified {
+            return false;
+        };
+
+        let paths_differ_by_at_most_one = self
+            .old_proof
+            .preimage_path
+            .iter()
+            .zip(&self.new_proof.preimage_path)
+            .all(|(a, b)| {
+                if a == b {
+                    // If the remaining trees are identical, then only one needs to be verified.
+                    // This is purely an evaluation-time optimization. Don't try to reproduce in the circuit, which cannot shortcut.
+                    return old_verified;
+                }
+                let differing_position_count = a
+                    .iter()
+                    .zip(b)
+                    .map(|(x, y)| x != y)
+                    .fold(0, |acc, are_diff| acc + are_diff as usize);
+                differing_position_count <= 1
+            });
+
+        // This is purely an evaluation-time optimization. Don't try to reproduce in the circuit, which cannot shortcut.
+        if !paths_differ_by_at_most_one {
+            return false;
+        };
+
+        self.new_proof.verify(new_root, key, new_value, hash_cache)
+    }
+}
+
+impl<'a, F: LurkField, const ARITY: usize, const HEIGHT: usize> Trie<'a, F, ARITY, HEIGHT> {
+    fn compute_hash(hash_cache: &PoseidonCache<F>, preimage: [F; ARITY]) -> F {
+        hash_cache.compute_hash(preimage)
+    }
+
+    pub fn root(&self) -> F {
+        self.root
+    }
+
+    /// How many leaves does this `Trie` have?
+    pub const fn leaves(&self) -> usize {
+        self.row_size(HEIGHT)
+    }
+
+    /// How many nodes does the `row`th row have?
+    pub const fn row_size(&self, row: usize) -> usize {
+        debug_assert!(row <= HEIGHT);
+        (ARITY as u32).pow(row as u32) as usize
+    }
+
+    fn register_hash(&mut self, preimage: [F; ARITY]) -> F {
+        let hash = Self::compute_hash(self.hash_cache, preimage);
+
+        self.children.insert(FWrap(hash), preimage);
+        hash
+    }
+
+    fn get_hash_preimage(children: &ChildMap<F, ARITY>, hash: F) -> Option<&[F; ARITY]> {
+        children.get(&FWrap(hash))
+    }
+
+    // TODO: This is expensive and can be cached for the type.
+    fn init_empty(&mut self) {
+        self.empty_roots = [F::zero(); HEIGHT];
+
+        if HEIGHT == 0 {
+            return;
+        };
+
+        let mut preimage = [self.empty_roots[0]; ARITY];
+
+        for i in 0..HEIGHT {
+            let hash = self.register_hash(preimage);
+
+            self.empty_roots[i] = hash;
+
+            preimage = [hash; ARITY];
+            for elt in preimage.iter_mut().take(ARITY) {
+                *elt = hash;
+            }
```

Comment:
I'm not sure what you mean, but I pushed a commit that tries to make what's happening a *little* more obvious.

Basically, at each row, starting from the bottom, we fill a preimage with repeated elements of the 'empty root' for the previous (lower) row. Then we hash that uniformly empty preimage to get the 'empty root' for the current row.

With that context, what is being done twice but shouldn't be?

Reply:
The lines that this commented on were:
```rust
            preimage = [hash; ARITY];
            for elt in preimage.iter_mut().take(ARITY) {
                *elt = hash;
            }
```
The things that seem to be done twice are, on the one hand:
```rust
            preimage = [hash; ARITY];
```
and in the other hand:
```rust
            for elt in preimage.iter_mut().take(ARITY) {
                *elt = hash;
            }
```

Reply:
Oh yes, I see what you mean. You are correct. Fixed.

Reply:
I pushed a second fix that takes the verbose version. I added that to avoid the extra (maybe?) allocation and forgot to remove the original afterward. Or… that's what I think happened.